opening_hours/utils/time_span_element.py

```diff
@@ -0,0 +1,364 @@
+        --------|--------------|--------   # No
+                |              |           #
+                |--------------|--------   # Yes
+                |            --|--------   # Yes
+                |              |--------   # No
+                |              |  ------   # No
+        """
+        return other.buffered_start_datetime <= self.buffered_start_datetime < other.buffered_end_datetime
+
+    def ends_inside_of(self, other: "TimeSpanElement") -> bool:
+        """
+        Does this time spans end inside the other time span?
+
+                     other
+                 <--timespan-->
+        ------  |              |           # No
+        --------|              |           # No
+        --------|--            |           # Yes
+        --------|--------------|           # Yes
+                |              |           #
+                |  ----------  |           # Yes
+                |--------------|           # Yes
+        --------|--------------|--------   # No
+                |              |           #
+                |--------------|--------   # No
+                |            --|--------   # No
+                |              |--------   # No
+                |              |  ------   # No
+        """
+        return other.buffered_start_datetime < self.buffered_end_datetime <= other.buffered_end_datetime
+
+    def can_fit_reservation_for_reservation_unit(
+        self,
+        reservation_unit: "ReservationUnit",
+        minimum_duration_minutes: int,
+    ) -> bool:
+        """Is this timespan long enough for a reservation for the given ReservationUnit?"""
+        # Minimum duration of a reservation for this ReservationUnit
+        reservation_unit_minimum_duration_minutes = (
+            minimum_duration_minutes
+            if not reservation_unit.min_reservation_duration
+            else max(reservation_unit.min_reservation_duration.total_seconds() / 60, minimum_duration_minutes)
+        )
+
+        front_buffered_reservation_unit_minimum_duration_minutes = reservation_unit_minimum_duration_minutes
+        back_buffered_reservation_unit_minimum_duration_minutes = reservation_unit_minimum_duration_minutes
+        if reservation_unit.buffer_time_before is not None:
+            front_buffered_reservation_unit_minimum_duration_minutes += (
+                reservation_unit.buffer_time_before.total_seconds() / 60
+            )
+        if reservation_unit.buffer_time_after is not None:
+            back_buffered_reservation_unit_minimum_duration_minutes += (
+                reservation_unit.buffer_time_after.total_seconds() / 60
+            )
+
+        return (
+            self.duration_minutes >= reservation_unit_minimum_duration_minutes
+            and self.front_buffered_duration_minutes >= front_buffered_reservation_unit_minimum_duration_minutes
+            and self.back_buffered_duration_minutes >= back_buffered_reservation_unit_minimum_duration_minutes
```

Comment:
?Should there nevertheless be a 4th condition: `... and self.buffered_duration_minutes >= buffered_reservation_unit_minimum_duration_minutes`

Maybe this possibility is taken care of otherwise but even so for this logic to be self-contained entity:
- time span is 90 minutes with no buffers before or after
- min. res. duration is 30 minutes but res. unit has 60 min. buffer settings both before and after
- thus (90 >= 30 and 90 >= [90 = 60 + 30] and 90 >= [90 = 30 + 60]) = true, but the 90 min slot cannot fit the 60 + 30 + 60 = **150** total duration with buffers -- unless the slot boundaries are defined purely by opening hours, not other reservations

Furthermore: is it so that we are now not excluding those kind of slots where self.front/back_buffered... = self.duration_minutes because nominally there is no buffer (determined by existing reservation before or after) but actually there is buffer (before and/or after), in practice, when there is (hard) closed/non-reservable time, over which the buffers of the hypothetical reservation could be plaed?

Reply:
Yes, fixed this and added test.